coveo-testing/README.md

```diff
@@ -221,3 +221,456 @@ If you run `pytest --collect-only` you will obtain the following:
+
+```python
+with patch(*ref(my_function)): ...
+with patch(*ref(my_function, context=...)): ...
+with patch(*ref(MyClass)): ...
+with patch(*ref(MyClass, context=...)): ...
+```
+
+The `context` is a slightly confusing topic that is explained in the `Demo` above.
+
+### Mock a method on a class
+
+Since a method cannot be imported and can only be accessed through the use of a dot `.` on a class or instance, 
+you can always patch methods globally:
+
+```python
+with patch(*ref(MyClass.fn)): ...
+```
+
+This is because no module can import `fn`; it has to go through an import of `MyClass`.
+
+### Mock a method on one instance of a class
+
+Simply add `obj=True` and use `patch.object()`:
+
+```python
+with patch.object(*ref(instance.fn, obj=True)): ...
+```
+
+
+### Mock an attribute on a class/instance/module/function/object/etc
+
+`ref` cannot help with this task:
+- You cannot refer an attribute that exists (you would pass the value, not a reference)
+- You cannot refer an attribute that doesn't exist (because it doesn't exist!)
+
+For this, there's no going around hardcoding the attribute name in a string:
+
+```python
+class MyClass:
+    def __init__(self) -> None:
+        self.a = 1
+
+
+def test_attr() -> None:
+    instance = MyClass()
+    with patch.object(instance, "a", new=2):
+        assert instance.a == 2
+        assert MyClass().a == 1
+```
+
+This sometimes work when patching **instances**. 
+The example works because `a` is a simple attribute that lives in `instance.__dict__` and `patch.object` knows
+about that.
+
+But if you tried to patch `MyClass` instead of `instance`, `mock.patch` would complain that there's no 
+such thing as `a` over there.
+Thus, patching an attribute globally will most likely result in a lot of wasted time, and should be avoided.
+
+There's no way to make the example work with `ref` because there's no way to refer `instance.a` without actually
+getting the value of `a`, unless we hardcode a string, which defeats the purpose of `ref` completely.
+
+
+### Mock a property
+
+You can only patch a property globally, through its class:
+
+```python
+class MyClass:
+    @property
+    def get(self) -> bool:
+        return False
+```
+
+```python
+from unittest.mock import PropertyMock, patch, MagicMock
+from coveo_testing.mocks import ref
+
+from mymodule import MyClass
+
+@patch(*ref(MyClass.get), new_callable=PropertyMock, return_value=True)
+def test(my_class_get_mock: MagicMock) -> None:
+    assert MyClass().get == True
+    my_class_get_mock.assert_called_once()
+```
+
+You **cannot** patch a property on an instance, this is a limitation of `unittest.mock` because of the way
+properties work.
+If you try, `mock.patch.object()` will complain that the property is read only.
+
+
+### Mock a classmethod or staticmethod on a specific instance
+
+When inspecting these special methods on an instance, `ref` ends up finding the class; the instance is missing.

```

Comment:
Maybe it's just me but this sentence is throwing me off. I think that it's the `;` which adds confusion. I'm not 100% sure what you're trying to say with that sentence. Is it "because ref finds the class, the instance is missing" or maybe "ref ends up finding the class because the instance is missing"?

Reply:
rephrased: `When inspecting these special methods on an instance, `ref` ends up finding the class instead of the instance.`